contracts/instruments/StakedPut.sol

```diff
@@ -0,0 +1,326 @@
+            instrumentPositions[account][positionID];
+
+        if (position.exercised) return 0;
+
+        uint256 profit = 0;
+
+        uint256 amount = position.amount;
+
+        uint256 strikePrice;
+        uint32 optionID;
+
+        strikePrice = position.putStrikePrice;
+        optionID = position.putOptionID;
+
+        address adapterAddress = factory.getAdapter(venue);
+        require(adapterAddress != address(0), "Adapter does not exist");
+
+        bool exercisable = adapter.canExercise(optionsAddress, optionID, amount);
+        if (!exercisable) {
+            return 0;
+        }
+
+        profit += adapter.delegateExerciseProfit(optionsAddress, optionID, amount);
+        return profit;
+    }
+
+    function canExercise(address account, uint256 positionID)
+        external
+        view
+        returns (bool)
+    {
+        InstrumentPosition storage position =
+            instrumentPositions[account][positionID];
+
+        if (position.exercised) return false;
+
+        bool canExercisePut = false;
+
+        uint256 strikePrice;
+        uint32 optionID;
+
+        strikePrice = position.putStrikePrice;
+        optionID = position.putOptionID;
+
+        address adapterAddress = factory.getAdapter(venue);
+        require(adapterAddress != address(0), "Adapter does not exist");
+
+        bool canExerciseOptions =
+            adapter.canExercise(optionsAddress, optionID, position.amount);
+
+        if (canExerciseOptions) {
+            canExercisePut = true;
+        }
+        return canExercisePut;
+    }
+
+//make this internal for production
+    function buyPutFromAdapter(BuyInstrumentParams calldata params)
+       public
+        payable
+        nonReentrant
+        returns (uint256 positionID)
+    {
+        require(block.timestamp < params.expiry, "Cannot purchase after expiry");
+
+        address adapterAddress = factory.getAdapter(venue);
+        require(adapterAddress != address(0), "Adapter does not exist");
+
+        ProtocolAdapterTypes.OptionTerms memory optionTerms = ProtocolAdapterTypes.OptionTerms(
+            underlying,
+            strikeAsset,
+            collateralAsset,
+            params.expiry,
+            params.putStrikePrice,
+            optionType,
+            params.paymentToken
+        );
+
+        uint256 optionID256 = adapter.delegatePurchase(optionTerms, params.optionAmount, params.putMaxCost);
+        uint32 optionID = uint32(optionID256);
+
+        InstrumentPosition memory position =
+            InstrumentPosition(
+                false,
+                venueID,
+                optionID,
+                params.optionAmount,
+                params.putStrikePrice,
+                params.expiry
+            );
+
+        positionID = instrumentPositions[msg.sender].length;
+        instrumentPositions[msg.sender].push(position);
+
+        uint256 balance = address(this).balance;
+        if (balance > 0) payable(msg.sender).transfer(balance);

```

Comment:
We should move these lines to the parent function `buyInstrument` so that we refund the change to the user for both buying the LP and `buyPutFromAdapter`